File: Quest/src/EffectModel.cpp
#include "include/EffectModel.hpp"
#include "include/Logger.hpp"
#include "include/Config.hpp"

bool Weather::EffectModel::GetEffectEnabledByName(Il2CppString* name)
{
    std::string key = GetNameWithoutSceneName(name);
    bool found = false;
    for (auto const &e: getConfig().config["enabledEffects"].GetArray()) {
        GenericLogger::Log( e.GetString());
        if ( e.GetString() == key) {
            found = true;
            std::string e2 = e.GetString();
            GenericLogger::Log("Found! " + e2 + " " + key);
            break;
        }
    }
    return found;
}
bool hasEnding (std::string fullString, std::string ending) {
    if (fullString.length() >= ending.length()) {
        return (0 == fullString.compare (fullString.length() - ending.length(), ending.length(), ending));
    } else {
        return false;
    }
}

bool Weather::EffectModel::IsEffectSeperateType(std::string name)
{
    return hasEnding(name, "Menu") || hasEnding(name, "Game");
}

bool Weather::EffectModel::IsEffectSeperateType(Il2CppString* name)
{
    std::string str = to_utf8(csstrtostr(name));
    return IsEffectSeperateType(str);
}

std::string Weather::EffectModel::GetNameWithoutSceneName(std::string name)
{
    if(IsEffectSeperateType(name))
    {
        return name.substr(0, name.length()-4);
    }
    return name;
}

std::string Weather::EffectModel::GetNameWithoutSceneName(Il2CppString* name)
{
    std::string convert = to_utf8(csstrtostr(name));
    return GetNameWithoutSceneName(convert);
}

/*void Weather::EffectModel::EnableEffect(Il2CppString* name, bool Value)
{
    std::string NewName = GetNameWithoutSceneName(name);
    std::string Game = NewName + "Game";
    std::string Menu = NewName + "Menu";
    if(IsEffectSeperateType(name))
    {
        Weather::Effect* effGame = GetEffectByName(Game);
        Weather::Effect* effMenu = GetEffectByName(Menu);
        effGame->enabled = Value;
        effMenu->enabled = Value;
        
        if (Value)
        {
            Config.enabledEffects.insert(NewName);
            getConfig().Write();
        }
        else
        {
            Config.enabledEffects.erase(NewName);
            getConfig().Write();
        }
        
        effGame.SetActiveRefs();
        effMenu.SetActiveRefs();

        return;
    }

}*/

File: PC/EffectModel.cs
namespace Weather
{
    internal static class EffectModel
    {
        public static Effect GetEffectByName(string name)
        {
            Effect @out = null;
            foreach (var e in BundleLoader.Effects)
            {
                if (e.Desc.effectName == name) @out = e;
            }

            return @out;
        }

        public static bool GetEffectEnabledByName(string name)
        {
            return PluginConfig.Instance.EnabledEffects.Contains(GetNameWithoutSceneName(name));
        }

        //Idk what to name this but basically this says if it should be seperated and have Game/Menu removed
        private static bool IsEffectSeparateType(string name)
        {
            if (name == null) return false;
            Plugin.Log.Debug("{IsEffectSeparateType} " + name + (name.EndsWith("Menu") || name.EndsWith("Game")) );
            return name.EndsWith("Menu") || name.EndsWith("Game");
        }

        public static bool IsEffectSeparateType(this Effect eff) => IsEffectSeparateType(eff.Desc.effectName);

        //Idk what to name this either but it gets the name without scene separation so RainMenu->Rain or RainGame->Rain
        public static string GetNameWithoutSceneName(string name)
        {
            if (!IsEffectSeparateType(name)) return name;

            //Game
            var @new = name.Substring(0, name.Length-4);
            return @new;
        }

        public static string GetNameWithoutSceneName(this Effect eff) => GetNameWithoutSceneName(eff.Desc.effectName);

        public static void EnableEffect(string name, bool value)
        {
            var newName = GetNameWithoutSceneName(name);
            var game = newName + "Game";
            var menu = newName + "Menu";

            if (IsEffectSeparateType(name))
            {
                Plugin.Log.Info(newName + " Is Independent " + name);
                var effGame = GetEffectByName(game);
                var effMenu = GetEffectByName(menu);
                effGame.Enabled = value;
                effMenu.Enabled = value;

                if (value)
                {
                    PluginConfig.Instance.AddEffect(newName);
                }
                else
                {
                    PluginConfig.Instance.EnabledEffects.Remove(newName);
                }
                
                effGame.SetActiveRefs();
                effMenu.SetActiveRefs();

                return;
            }

            var eff = GetEffectByName(name);
            if (eff == null) return;
            eff.Enabled = value;

            if (value)
            {
                PluginConfig.Instance.AddEffect(name);
            }
            else
            {
                PluginConfig.Instance.EnabledEffects.Remove(name);
            }

            eff.SetActiveRefs();
        }
    }
}
